### PR TITLE
Additional Daemon dns and dhcp

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Enable dns/dhcp daemon on a specific network interface (like portal) 

# Impacts
DHCP/DNS

# Delete branch after merge
YES

# Checklist
- [n/a] Document the feature
- [n/a] Add unit tests
- [n/a] Add acceptance tests (TestLink)

# NEWS file entries
## New Features
* DHCP and DNS services can be enabled on a specific interface.

### CODE REVIEW

File: go/coredns/plugin/pfdns/pfdns.go [hunk 1]
```diff
@@ -352,7 +355,7 @@ func (pf *pfdns) detectVIP() error {
 	var keyConfCluster pfconfigdriver.NetInterface
 	keyConfCluster.PfconfigNS = "config::Pf(CLUSTER," + pfconfigdriver.FindClusterName(ctx) + ")"
 
-	for _, v := range interfaces.Element {
+	for _, v := range append(interfaces.Element, DNSinterfaces.Element...) {
```
Comment (julsemaan) [inline go/coredns/plugin/pfdns/pfdns.go]: Will this be unique ?

File: go/dhcp/config.go [hunk 2]
```diff
@@ -62,14 +62,17 @@ func (d *Interfaces) readConfig() {
...
+
 	var keyConfNet pfconfigdriver.PfconfigKeys
 	keyConfNet.PfconfigNS = "config::Network"
 	keyConfNet.PfconfigHostnameOverlay = "yes"
 
 	pfconfigdriver.FetchDecodeSocket(ctx, &keyConfNet)
 
 	wg := &sync.WaitGroup{}
-	for _, v := range interfaces.Element {
+	for _, v := range append(interfaces.Element, DHCPinterfaces.Element...) {
```
Comment (julsemaan) [inline go/dhcp/config.go]: Will this be unique ?

File: go/dhcp/main.go [hunk 3]
```diff
@@ -82,10 +82,12 @@ func main() {
 	VIPIp = make(map[string]net.IP)
 
 	go func() {
+		var DHCPinterfaces pfconfigdriver.DHCPInts
+		pfconfigdriver.FetchDecodeSocket(ctx, &DHCPinterfaces)
 		var interfaces pfconfigdriver.ListenInts
 		pfconfigdriver.FetchDecodeSocket(ctx, &interfaces)
 		for {
-			DHCPConfig.detectVIP(interfaces)
+			DHCPConfig.detectVIP(append(interfaces.Element, DHCPinterfaces.Element...))
```
Comment (julsemaan) [inline go/dhcp/main.go]: Will this be unique ?

**Comment by fdurand** [comment]
Bump

File: conf/documentation.conf [hunk 4]
```diff
@@ -763,6 +763,8 @@ Describes type of the named interface.
  * "high-availability" is for an interface between two PacketFence servers dedicated to high-availability (drbd, corosync).
  * "portal" interface have the captive-portal running on them without the need of having any enforcement mechanism.
  * "radius" interface have the radius running (auth, acct, cli)
+ * "dhcpd" interface have the pfdhcp running.
```
Comment (julsemaan) [inline conf/documentation.conf]: Like we don't call the radius interface radiusd, this should be called dhcp not dhcpd

Comment (julsemaan) [reply conf/documentation.conf]: And it should be the "pfdhcp daemon"

File: conf/documentation.conf [hunk 5]
```diff
@@ -763,6 +763,8 @@ Describes type of the named interface.
...
+ * "dns" interface have the pfdns running
```
Comment (julsemaan) [inline conf/documentation.conf]: "pfdns daemon"

File: go/sharedutils/util.go [hunk 6]
```diff
@@ -267,3 +267,22 @@ func CleanIP(s string) (string, error) {
...
 	}
 }
+
+func RemoveDuplicates(elements []string) []string {
+	// Use map to record duplicates as we find them.
+	encountered := map[string]bool{}
+	result := []string{}
+
+	for v := range elements {
+		if encountered[elements[v]] == true {
```
Comment (julsemaan) [inline go/sharedutils/util.go]: Could just negate this and move the content of the else in here instead of having an empty if body

**Comment by fdurand** [comment]
Bump

**Comment by nqb** [comment]
With this feature, will it be possible to enable routed networks on interfaces where you enable additional DHCP and DNS daemons ?

**Comment by fdurand** [comment]
BUMP

**Comment by fdurand** [comment]
Change has been made in the new gui for this PR.

**Pull request merged**